### DISCUSSION TITLE
cross-compiling fprime for BeagleBone

### DISCUSSION THREAD

**Discussion by capsulecorplab** [nasa/fprime#1437] [General]
How difficult would it be to write a cmake build system for cross-compiling fprime for a BeagleBone?

**Comment by SterlingPeet** [comment]
Seems like it shouldn’t be too crazy, some combination of the Raspberry Pi example and something like this link should get you there:

https://parzival2.github.io/blog/posts/beaglebone-cross-compile/

I would be interested in this myself because it would be a nice example for training my students, but lately the externally funded projects have been too far away from BeagleBone for my to take a serious look at it.

Thanks,

Sterling Peet

Sent from my mobile.

On May 6, 2022, at 2:23 PM, Sean Marquez ***@***.***> wrote:

﻿

How difficult would it be to write a cmake build system for cross-compiling fprime for a BeagleBone?

**Reply by LeStarch** [reply]
Looks like all the beagle bones are 32-bit ARM processors. Thus you'd use the Linux + 32-bit arm compiler and set it up to look much like the raspberrypi toolchain.

**Comment by capsulecorplab** [comment]
Assuming I manage to write a cross compiler for the beaglebone, how might I go about connecting to, say, the Beaglebone Black (which doesn't have a WiFi module) via `fprime-gds`?

**Reply by timcanham** [reply]
The BeagleBone Black has a wired Ethernet connection, or you could use one of the UARTs.

**Reply by SterlingPeet** [reply]
I am actually using a USB wifi adapter, it seems to work just fine for a demo.

**Comment by capsulecorplab** [comment]
I managed to write a cmake build system for cross-compiling fprime for the BeagleBone! It's actually more slimmed down than the one for the `raspberrypi`. I tested it by building and porting the reference implementation + math components to a BeagleBone Blue and performing math operations with the `MathSender` component via `fprime-gds` connected via TCP. This is my first attempt at writing a cmake file, so feel free to suggest any improvements...

Branch with just the cmake file for BeagleBone:
https://github.com/capsulecorplab/fprime/tree/cross-compile-for-beaglebone-v3-0-0

Branch with the cmake file for BeagleBone + MathComponent:
https://github.com/capsulecorplab/fprime/tree/cross-compile-for-beaglebone-with-fprime-v3.0.0

**Reply by SterlingPeet** [reply]
Thats great!  I can confirm that I have cross compiled Ref for beaglebone using an parallel strategy to get there, compiling on Ubuntu Focal.  I am setting up to give a lecture and have students compile a demo using my cross-compile setup next monday, so I'll be cleaning up my solutions and posting them to github afterward.

**Comment by capsulecorplab** [comment]
Is there a preferred workflow for setting up the project scaffolding for an fprime project? I'm thinking of creating an example fprime project, similar to [fprime-arduino](https://github.com/fprime-community/fprime-arduino) or the `Ref` & `RPI` examples in the `fprime` repo, that would include components s.a., the blinking LED and Math components, but for the BeagleBone. If of interest, I'm also willing to transfer the ownership of the project/repo over to https://github.com/fprime-community/

**Reply by LeStarch** [reply]
We are working this summer on a solid reference example that will provided exactly this.  Typically, we pull the toolchain and CMake support files into a library style module and the include that in an example project.

**Reply by capsulecorplab** [reply]
Would this happen to be the system reference? I'd like to start experimenting on other platforms, such as the ESP32/8266 (see #1612) or possibly RPI Pico, as well as clean up my BeagleBone fork, following expected patterns for developing example projects and libraries, so any guidance for how to do so would be appreciated

**Comment by SterlingPeet** [comment]
I actually got BeagleBone as a target platform working using vagrant/virtualbox last fall, and had a class full of students cross-compiling a deployment for beaglebone.  It took few hours to sort through at the time, but it is very doable: just a linux target from the F Prime perspective.

**Reply by SterlingPeet** [reply]
(hah!  see an earlier reply for details.)

**Reply by capsulecorplab** [reply]
@SterlingPeet Speaking of virtual environments, let me know if you'd like to try out my immutable-infrastructure-as-code (IIaC) workspace for fprime https://github.com/capsulecorplab/fprime-workspace-image 😉

**Reply by SterlingPeet** [reply]
I better put a pin in that, I'm probably going to be pretty sparse on F Prime work until June. 😬

**Comment by LeStarch** [comment]
Note: we've gotten Beaglebone working using: https://github.com/fprime-community/fprime-arm-linux and a prebuilt ARM toolchain, e.g. Linaro 7.5, or possibly ARM 10.2 from either of those websites.
